blockchain/validators/src/metadata_key.rs

```diff
@@ -0,0 +1,131 @@
+        for config in configs {
+            keys.entry(config.message_signing_pub_key)
+                .or_default()
+                .insert(config.clone());
+        }
+
+        Self { keys }
+    }
+
+    /// Check whether the given block metadata is valid at the given index.
+    pub fn verify(&self, metadata: &BlockMetadata, block_index: BlockIndex) -> Result<(), Error> {
+        let ranges = self
+            .keys
+            .get(metadata.node_key())
+            .ok_or(Error::UnknownPubKey)?;
+
+        for range in ranges {
+            if range.index_in_range(block_index) {
+                return Ok(metadata.verify()?);
+            }
+        }
+
+        Err(Error::InvalidPubKey)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mc_blockchain_test_utils::{make_block_id, make_block_metadata_contents};
+    use mc_crypto_keys::{Ed25519Pair, Ed25519Signature};
+    use mc_util_from_random::FromRandom;
+    use mc_util_test_helper::{RngType as FixedRng, SeedableRng};
+
+    fn make_key(seed: u64) -> Ed25519Public {
+        *make_metadata(seed).node_key()
+    }
+
+    fn make_metadata(seed: u64) -> BlockMetadata {
+        let seed_bytes = seed.to_be_bytes();
+        let mut seed = [0u8; 32];
+        seed[0..seed_bytes.len()].copy_from_slice(&seed_bytes);
+        let mut rng = FixedRng::from_seed(seed);
+
+        let signer = Ed25519Pair::from_random(&mut rng);
+
+        BlockMetadata::from_contents_and_keypair(
+            make_block_metadata_contents(make_block_id(&mut rng), &mut rng),
+            &signer,
+        )
+        .expect("BlockMetadata::from_contents_and_keypair")
+    }
+
+    fn make_verifier() -> MetadataKeyValidator {
+        MetadataKeyValidator::new(&[
+            NodeKeyValidity::new(make_key(1), 0, None),
+            NodeKeyValidity::new(make_key(2), 0, 10),
+            NodeKeyValidity::new(make_key(3), 10, None),
+        ])
+    }
+
+    #[test]
+    fn happy_path() {
+        let verifier = make_verifier();
+        let metadata = make_metadata(1);
+        assert_eq!(verifier.verify(&metadata, 0), Ok(()));
+        assert_eq!(verifier.verify(&metadata, 10), Ok(()));
+        assert_eq!(verifier.verify(&metadata, 100), Ok(()));
+        assert_eq!(verifier.verify(&metadata, 1000), Ok(()));
+    }
+
+    #[test]
+    fn unrecognized_key() {
+        let verifier = make_verifier();
+        let metadata = make_metadata(10);
+        assert_eq!(verifier.verify(&metadata, 0), Err(Error::UnknownPubKey));
+    }
+
+    #[test]
+    fn expired_key() {
+        let verifier = make_verifier();
+        let metadata = make_metadata(2);
+        assert_eq!(verifier.verify(&metadata, 11), Err(Error::InvalidPubKey));
+        assert_eq!(verifier.verify(&metadata, 10), Ok(()));
+    }
+
+    #[test]
+    fn premature_key() {
+        let verifier = make_verifier();
+        let metadata = make_metadata(3);
+        assert_eq!(verifier.verify(&metadata, 1), Err(Error::InvalidPubKey));
```

Comment:
Should test right at the boundary to avoid off by ones
```suggestion
        assert_eq!(verifier.verify(&metadata, 9), Err(Error::InvalidPubKey));
```